### PR TITLE
Userland: Add a simple mktemp(1) utility

### PR DESCRIPTION
**Pull request by IdanHo**
This was needed for a port

### CODE REVIEW

File: Userland/Utilities/mktemp.cpp [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    for (auto i = pattern_length - 1; i >= 0; --i) {
+        if (pattern[i] != 'X')
+            break;
+        new_filename[i] = random_characters[(arc4random() % (sizeof(random_characters) - 1))];
+    }
+
+    return new_filename;
+}
+
+static const char* make_temp(const char* pattern, bool directory, bool dry_run)
```
Comment (ADKaster) [inline Userland/Utilities/mktemp.cpp:51]: Is there anything special about the `mktemp(1)` utility that means you can't use `mktemp(3)` and `mkdtemp` from stdlib.h? I guess maybe the dry run option?

File: Userland/Utilities/mktemp.cpp [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+
+constexpr const char* default_template = "tmp.XXXXXXXXXX";
+
+static const char* generate_random_filename(const char* pattern)
+{
+    char* new_filename = strdup(pattern);
```
Comment (ADKaster) [inline Userland/Utilities/mktemp.cpp:38]: This string is never freed

File: Userland/Utilities/mktemp.cpp (hunk 1 above)
Comment (bcoles) [reply Userland/Utilities/mktemp.cpp:51]: > 
> 
> Is there anything special about the `mktemp(1)` utility that means you can't use `mktemp(3)` and `mkdtemp` from stdlib.h? I guess maybe the dry run option?

`mktemp` can be used to perform a dry run. It does not create the associated file.

File: Userland/Utilities/mktemp.cpp [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+#include <AK/LexicalPath.h>
+#include <LibCore/ArgsParser.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+
+constexpr const char* default_template = "tmp.XXXXXXXXXX";
+
+static const char* generate_random_filename(const char* pattern)
```
Comment (bcoles) [inline Userland/Utilities/mktemp.cpp:36]: `stdlib` also includes `generate_unique_filename` for this purpose.

**Comment by bgianfo** [comment]
Consider adding a man page for your nice new shiny utility! 👍

File: Userland/Utilities/mktemp.cpp (hunk 1 above)
Comment (IdanHo) [reply Userland/Utilities/mktemp.cpp:51]: mktemp(3) only allows templates with exactly 6 Xs (6 random characters), while mktemp(1) is specified to allow any amount larger than 3, but yes the dry run option as well

File: Userland/Utilities/mktemp.cpp (hunk 3 above)
Comment (IdanHo) [reply Userland/Utilities/mktemp.cpp:36]: same problem as using mktemp(3) directly, mktemp(3) only allows templates with exactly 6 Xs (6 random characters), and as a result the implementation of generate_unique_filename in stdlib does as well, while mktemp(1) is specified to allow any amount larger than 3

File: Userland/Utilities/mktemp.cpp (hunk 2 above)
Comment (IdanHo) [reply Userland/Utilities/mktemp.cpp:37]: oops! fixed

**Pull request merged**